### PR TITLE
i122 - Get web app working

### LINKED ISSUE #122: Bring the web app up to parity with the mobile apps
Currently we have a separate entry point for web (`lib/main_web.dart`). This was mostly so we could get a build up easily and have an accessible privacy policy.  Having a separate entry point is also a nice way of having different design and behaviour on the web version.  We could of course have a single entry point and add platform checks to get different behaviour, personally I think this is more clunky. 

The web app uses the same `pubspec.yaml` and `Firestore`, `FirebaseAuth`, etc have been setup to build (and should be fine to use in the app).  Anything that's not web compatible we can disregard via conditional compilation: 
[Flutter: Conditional Compilation for Web - gskinner blog](https://blog.gskinner.com/archives/2020/03/flutter-conditional-compilation-for-web.html)
  - probably easier to use [platform](https://pub.dev/packages/platform) than the way they do it 

No web support (not a complete list yet...) 
- Firebase Messaging, see [[firebase_messaging] Add web support · Issue #45300 · flutter/flutter](https://github.com/flutter/flutter/issues/45300)
- ~~flutter_facebook_login (there is an issue up to replace facebook with apple sign in)~~ (EDIT: removed plugin) 
- ~~[apple_sign_in | Flutter Package](https://pub.dev/packages/apple_sign_in)~~ (EDIT: replaced with sign_in_with_apple) 

If we want to write a web plugin in place of one that doesn't support web: 
- [How to Write a Flutter Web Plugin - Flutter - Medium](https://medium.com/flutter/how-to-write-a-flutter-web-plugin-5e26c689ea1) 
- [How To Write a Flutter Web Plugin: Part 2 - Flutter - Medium](https://medium.com/flutter/how-to-write-a-flutter-web-plugin-part-2-afdddb69ece6) 

If you are interested, [Web Plugins](https://github.com/flutter/flutter/projects/69) tracks the progress of web support for all the firebase plugins.

### LINKED ISSUE #164: Get Sign In With Apple working for web app
#122 is has google auth working on the web app (no work was needed as the google_sign_in plugin works out of the box on web) 

`sign_in_with_apple` does not work for web out of the box, but they recommend a way to do it and it plan to add it at some point (though not soon), see: https://github.com/aboutyou/dart_packages/issues/90#issuecomment-637580587

From that comment the steps would be: 

- [ ] render the button in Flutter 
- [ ] use the same URL-building logic as on Android, see: https://github.com/aboutyou/dart_packages/blob/23679ba7e4b60d17f5ebe69f142fe94d813b7df7/packages/sign_in_with_apple/lib/src/sign_in_with_apple.dart#L198
- [ ] on button click open that URL (redirect of the current page or pop-up/new tab (so main app doesn't have to reload on return)
- [ ] set a route on your app as the redirect URL

Further notes from the linked comment:
- the redirect URL would need to be able to handle the POST that Apple is sending to your servers 
- you can then either handle that directly on the server (and then for example pass down a cookie and the SPA entry page), or somehow pass received POST data to the client, so it would call the same API than the native apps do
- as Apple is always redirecting through the redirect URL, I don't yet know how to make that really nice from a client API perspective (so that the same simple method could be called on all stacks)
- maybe the pop-up / new tab approach can be made to work with the final redirect URL using postMessage to pass the data back to the Flutter app

### PR DESCRIPTION
**Pull request by nickmeinhold**
fixes #122 - Bring the web app up to parity with the mobile apps

- added `cdns_config.html` to .gitignore (workaround for running locally)
- added a launch configuration for running web
- updated firebase-functions package
- added checks to avoid plugins that don't support web
(firebase_messaging & sign_in_with_apple)
- changed main_web.dart to match main.dart (keeping as a separate
file as we will probably have different setups for web and mobile)

Note:
- running locally has problems, tracked in #165
- I made a wiki page to keep track of our approach for the web version
https://github.com/Adventures-In/chat_app/wiki/Web
- the code in this PR is live at: https://adventures-in.web.app

### CODE REVIEW

File: lib/auth_page.dart [hunk 1]
```diff
@@ -18,13 +19,14 @@ class AuthPage extends StatelessWidget {
...
             },
           ),
-          AppleSignInButton(
-              style: AppleButtonStyle.black,
-              onPressed: () async {
-                _appleSignin(context).listen((event) {
-                  print(event);
-                });
-              }),
+          if (!kIsWeb)
```
Comment (gaslitbytech) [inline lib/auth_page.dart:22]: Is there a limitation technical reason AppleSignIn will not work for web?

File: .vscode/launch.json [hunk 2]
```diff
@@ -9,6 +9,13 @@
             "request": "launch",
             "type": "dart"
         },
+        {
+            "name": "Web",
+            "request": "launch",
+            "type": "dart",
+            "program": "lib/main_web.dart",
+            "args": ["--web-port", "5000"]
```
Comment (gaslitbytech) [inline .vscode/launch.json:17]: I tried this config. It seems to run though http and https://127.0.0.1:5000 gets ERR_CONNECTION_REFUSED. Is there something I'm missing?

**Comment by nickmeinhold** [comment]
@tourdownunder sorry to be annoying but thought I would ping you to let you know this is ready to go through when/if approved. I'm pretty keen to add an admin console to the web app so we can do things like clean up orphaned User / Conversation objects. Thanks!

**Comment by gaslitbytech** [comment]
On mobile so brief. Not annoying though I tried to run it and wasn’t running for me as per earlier comment.

**Comment by nickmeinhold** [comment]
Thanks for the reply! Sorry which comment is that? Was it in a discussion?  There's no earlier comment from you in the PR that I can see...?

**Comment by nickmeinhold** [comment]
Oh I imagine it was due to #165 ?  Sorry my useless brain has forgotten we discussed you couldn't run locally.  Both problems (firebase config & whitelisting) have workarounds but they're both a bit dodgy and it'd be good to discuss I think.  I'll push out the current code to the hosting so you can see the web version running if you like.

**Comment by nickmeinhold** [comment]
Code from this (up to date) PR is live: https://adventures-in.web.app

**Review by gaslitbytech — APPROVED** [review]
Thanks for publishing this. 

I do now recall you mentioning the reason for not being able to run locally.

Again great work Nick

File: lib/auth_page.dart (hunk 1 above)
Comment (nickmeinhold) [reply lib/auth_page.dart:22]: Thanks for the question mate!  The PR got a bit unwieldy and has ended up a bit unclear sorry. Despite the branch being named `nickmeinhold/issue164`, the issue #164 is not going to be closed by this PR and will continue on for us to track auth for the web version. It has a bunch of info about Sign In With Apple on the web. To answer your question though, the plugin doesn't support web out of the box (and they don't plan to soon) so we'll need to roll our own.

File: .vscode/launch.json (hunk 2 above)
Comment (nickmeinhold) [reply .vscode/launch.json:17]: Yeah there are a couple of workarounds needed to get the app to run on web locally. Although I think in this case your were using https and the web server only runs http locally... maybe (?) 

Maybe we could discuss how we should approach running web locally when we next meet.

**Pull request merged**